### ISSUE TITLE
dual stack k3s kills default ::0 route after start

### ISSUE THREAD

**Issue by strayer** [k3s-io/k3s#5019]
**Environmental Info:**
K3s Version: v1.23.2+k3s1 - also tested on v1.23.2-rc1+k3s1 and  v1.21.9+k3s1

Node(s) CPU architecture, OS, and Version: 

Linux k3s-1 4.19.0-18-amd64 #1 SMP Debian 4.19.208-1 (2021-09-29) x86_64 GNU/Linux

(also tested with Debian 11)

Cluster Configuration:

Simple one node installation.

The Debian installation is unmodified, only `iface enp1s0 inet6 dhcp` is added to `/etc/network/interfaces` and `send dhcp-client-identifier = hardware;` to `/etc/dhcp/dhclient.conf` to receive a static IPv6 suffix by the DHCPv6 server.

The server receives a static IPv4 via DHCP (dnsmasq on OpenWRT) and a static IPv6 suffix via DHCPv6 (odhcpd on OpenWRT) with a ULA for static local addresses and a globally routable IPv6 with a dynamic prefix (enforced by ISP).

The server is a VM on an Unraid host running via KVM on a bridge to the local network. IPv6 works fine by itself locally and globally.

**Describe the bug:**

After starting k3s with `systemctl start k3s`, the default route for `::0` via enp1s0 is removed and replaced by a route via lo, breaking IPv6 on the system. Rebooting the system with k3s not on autostart fixes the issue until k3s is started again.

`route -6n` before starting k3s:

```
Destination                    Next Hop                   Flag Met Ref Use If
::1/128                        ::                         U    256 2     0 lo
2003:e5:d731:9d04::5/128       ::                         U    256 1     0 enp1s0
2003:e5:d731:9d04::/64         ::                         UAe  256 1     0 enp1s0
fd3b:2815:be50:4::5/128        ::                         U    256 1     0 enp1s0
fd3b:2815:be50:4::/64          ::                         UA   256 2     0 enp1s0
fe80::/64                      ::                         U    256 2     0 enp1s0
::/0                           fe80::dea6:32ff:fe1e:6d61  UGDAe 1024 2     0 enp1s0
::1/128                        ::                         Un   0   5     0 lo
2003:e5:d731:9d04::5/128       ::                         Un   0   4     0 enp1s0
2003:e5:d731:9d04:5054:ff:febf:d6e4/128 ::                         Un   0   2     0 enp1s0
fd3b:2815:be50:4::5/128        ::                         Un   0   3     0 enp1s0
fd3b:2815:be50:4:5054:ff:febf:d6e4/128 ::                         Un   0   3     0 enp1s0
fe80::5054:ff:febf:d6e4/128    ::                         Un   0   3     0 enp1s0
ff00::/8                       ::                         U    256 5     0 enp1s0
::/0                           ::                         !n   -1  1     0 lo
```

`route -6n` after starting k3s:

```
Destination                    Next Hop                   Flag Met Ref Use If
::1/128                        ::                         U    256 2     0 lo
2003:e5:d731:9d04::5/128       ::                         U    256 1     0 enp1s0
2003:e5:d731:9d04::/64         ::                         UAe  256 1     0 enp1s0
fc15:1::/128                   ::                         U    256 1     0 flannel-v6.1
fc15:1::/64                    ::                         U    256 1     0 cni0
fd3b:2815:be50:4::5/128        ::                         U    256 1     0 enp1s0
fd3b:2815:be50:4::/64          ::                         UA   256 6     0 enp1s0
fe80::/64                      ::                         U    256 2     0 enp1s0
fe80::/64                      ::                         U    256 1     0 flannel.1
fe80::/64                      ::                         U    256 1     0 flannel-v6.1
fe80::/64                      ::                         U    256 1     0 cni0
fe80::/64                      ::                         U    256 1     0 vethb3984b88
fe80::/64                      ::                         U    256 1     0 veth6a343a96
fe80::/64                      ::                         U    256 1     0 vethbff28e93
fe80::/64                      ::                         U    256 1     0 veth83596866
fe80::/64                      ::                         U    256 1     0 veth636b6c3c
::/0                           ::                         !n   -1  1     0 lo
::1/128                        ::                         Un   0   6     0 lo
2003:e5:d731:9d04::/128        ::                         Un   0   3     0 enp1s0
2003:e5:d731:9d04::5/128       ::                         Un   0   5     0 enp1s0
2003:e5:d731:9d04:5054:ff:febf:d6e4/128 ::                         Un   0   2     0 enp1s0
fc15:1::/128                   ::                         Un   0   3     0 flannel-v6.1
fc15:1::/128                   ::                         Un   0   3     0 cni0
fc15:1::1/128                  ::                         Un   0   2     0 cni0
fd3b:2815:be50:4::/128         ::                         Un   0   3     0 enp1s0
fd3b:2815:be50:4::5/128        ::                         Un   0   4     0 enp1s0
fd3b:2815:be50:4:5054:ff:febf:d6e4/128 ::                         Un   0   4     0 enp1s0
fe80::/128                     ::                         Un   0   5     0 enp1s0
fe80::/128                     ::                         Un   0   3     0 flannel-v6.1
fe80::/128                     ::                         Un   0   3     0 flannel.1
fe80::/128                     ::                         Un   0   3     0 veth636b6c3c
fe80::/128                     ::                         Un   0   3     0 cni0
fe80::/128                     ::                         Un   0   3     0 veth83596866
fe80::/128                     ::                         Un   0   3     0 vethbff28e93
fe80::/128                     ::                         Un   0   3     0 vethb3984b88
fe80::/128                     ::                         Un   0   3     0 veth6a343a96
fe80::f5:1cff:fe64:f6ec/128    ::                         Un   0   2     0 flannel-v6.1
fe80::2c3a:87ff:feb9:7beb/128  ::                         Un   0   2     0 vethb3984b88
fe80::4801:ccff:fed6:ab81/128  ::                         Un   0   2     0 veth636b6c3c
fe80::5054:ff:febf:d6e4/128    ::                         Un   0   7     0 enp1s0
fe80::6c85:8dff:fe50:c773/128  ::                         Un   0   3     0 flannel.1
fe80::74ee:aff:fe4a:7b56/128   ::                         Un   0   2     0 cni0
fe80::b4da:aff:fe03:b2a3/128   ::                         Un   0   2     0 vethbff28e93
fe80::c026:7cff:fe49:1adf/128  ::                         Un   0   4     0 veth83596866
fe80::d078:e8ff:fed4:eda6/128  ::                         Un   0   2     0 veth6a343a96
ff00::/8                       ::                         U    256 5     0 enp1s0
ff00::/8                       ::                         U    256 1     0 flannel.1
ff00::/8                       ::                         U    256 1     0 flannel-v6.1
ff00::/8                       ::                         U    256 4     0 cni0
ff00::/8                       ::                         U    256 1     0 vethb3984b88
ff00::/8                       ::                         U    256 1     0 veth6a343a96
ff00::/8                       ::                         U    256 1     0 vethbff28e93
ff00::/8                       ::                         U    256 1     0 veth83596866
ff00::/8                       ::                         U    256 1     0 veth636b6c3c
::/0                           ::                         !n   -1  1     0 lo
```

**Steps To Reproduce:**

* install default Debian 10 or 11
* configure system to use DHCPv6
* install k3s with:

```
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=v1.23.2+k3s1 sh -s - \
--disable-network-policy \
--node-ip="192.168.178.5,fd3b:2815:be50:4::5" \
--cluster-cidr="10.42.0.0/16,fc15:1::/56" \
--service-cidr="10.43.0.0/16,fc15:2::/112"
```

* run `route -6n` to inspect broken routing

**Expected behavior:**

The default IPv6 route should remain after starting k3s.

**Actual behavior:**

The default IPv6 route is replaced by a route via the loopback interface.

**Additional context / logs:**

ip addr output:
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp1s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 52:54:00:bf:d6:e4 brd ff:ff:ff:ff:ff:ff
    inet 192.168.178.5/24 brd 192.168.178.255 scope global dynamic enp1s0
       valid_lft 86367sec preferred_lft 86367sec
    inet6 fd3b:2815:be50:4::5/128 scope global 
       valid_lft forever preferred_lft forever
    inet6 2003:e5:d731:9d04::5/128 scope global 
       valid_lft forever preferred_lft forever
    inet6 fd3b:2815:be50:4:5054:ff:febf:d6e4/64 scope global dynamic mngtmpaddr 
       valid_lft forever preferred_lft forever
    inet6 2003:e5:d731:9d04:5054:ff:febf:d6e4/64 scope global dynamic mngtmpaddr 
       valid_lft 13739sec preferred_lft 1139sec
    inet6 fe80::5054:ff:febf:d6e4/64 scope link 
       valid_lft forever preferred_lft forever
```

**Backporting**

- [ ] Needs backporting to older releases

**Comment by strayer** [comment]
Update: when I'm not using DHCPv6 but SLAAC to get an IPv6 the default route does NOT get deleted. Not sure how this related, but it MAY be possible my local network setup is the culprit. Maybe someone with more knowledge about how Linux handles IPv6 addresses and routes can shine some light on this.

**Comment by brandond** [comment]
I'm not aware of anywhere in K3s, either in the kubelet/kube-proxy or Flannel CNI, that we add or remove host-level default routes. @manuelbuil can you think of anything on our side that might be causing this?

**Comment by manuelbuil** [comment]
> I'm not aware of anywhere in K3s, either in the kubelet/kube-proxy or Flannel CNI, that we add or remove host-level default routes. @manuelbuil can you think of anything on our side that might be causing this?

You are right, I don't think we are

**Comment by manuelbuil** [comment]
> Update: when I'm not using DHCPv6 but SLAAC to get an IPv6 the default route does NOT get deleted. Not sure how this related, but it MAY be possible my local network setup is the culprit. Maybe someone with more knowledge about how Linux handles IPv6 addresses and routes can shine some light on this.

Can you check the content of the DHCPv6 message you are getting? It might be messing up with your routing table. Although not sure how is k3s impacting :thinking:

**Comment by strayer** [comment]
Thank you both for looking into this! This may not be related to k3s at all. There was a configuration issue on the host system of the k3s VM related to Docker, libvirt and br_nftables. All of this might have been caused by this too.

I explained this in the Unraid forums, in case anyone wants some more context about this: https://forums.unraid.net/topic/118841-dhcpv6-not-working-for-vms-in-br0/?do=findComment&comment=1100864

I'm going to close this for now, since the k3s VM didn't have any problems for a while now.

**Comment by Moosemorals** [comment]
When k3s starts it sets the value of `/proc/sys/net/ipv6/conf/all/forwarding` to `1`. This means that the kernel stops paying attention to RA messages unless `/proc/sys/net/ipv6/conf/all/accept_ra` (and possibly `/proc/sys/net/ipv6/conf/{interface}/accept_ra`) is (are) set to `2`.
The practical result is that the default IPv6 route vanishes about 30 seconds after k3s starts.
See https://sysctl-explorer.net/net/ipv6/accept_ra/

**Comment by brandond** [comment]
I believe that this is by design, see https://github.com/kubernetes/kubernetes/issues/91507. As per that issue that I believe that enabling accept_ra by default when forwarding is enabled is considered unsafe.

@manuelbuil I seem to recall that there was some discussion about this on another issue, but I'm not finding it. Do you recall?

**Comment by Moosemorals** [comment]
You could remember the value of `accept_ra` and restore it after `forwarding` is updated?
If the original value is 0 (ignore ra) or one (accept ra, unless we're forwarding) then restoring it won't make any difference, and if it's 2 then someone must have set it to that explicitly.
